### ISSUE TITLE
Problem with nightly and rustflags "+atomics"

### ISSUE THREAD

**Issue by mhchia** [tlsnotary/tlsn-extension#6]
(Updated the instructions. Big thanks to @themighty1 for reproducing and finding the issues in these steps.)

## What's wrong?
To bring back rayon, we need [wasm-bindgen-rayon](https://github.com/GoogleChromeLabs/wasm-bindgen-rayon). It requires us to use nightly rust with rustflags for "+atomics,+bulk-memory,+mutable-globals"

## Steps to Reproduce
Step 1: Open a new terminal. Build and run a websocket proxy
```bash
git clone https://github.com/novnc/websockify && cd websockify
./docker/build.sh
docker run -it --rm -p 55688:80 novnc/websockify 80 twitter.com:443
```

Step 2: Open a new terminal. Run a modified notary server in twitter example
```bash
git clone git@github.com:mhchia/tlsn.git
cd tlsn && git checkout tlsn-examples-ws
cd tlsn/examples
RUST_LOG=debug,yamux=info cargo run --release --example notary
```

Step 3: Clone the branch `mhchia/debugging-atomics-issue` from tlsnotary/tlsn-extension. Using this branch to debug so that we don't need to load the extension.

```bash
git clone git@github.com:tlsnotary/tlsn-extension.git
cd tlsn-extension
git checkout mhchia/debugging-atomics-issue
```

Step 4: Use nightly rust by creating a file `rust-toolchain` with the content `nightly-2022-12-12` under tlsn-extension project root.
```bash
echo "nightly-2022-12-12" > rust-toolchain
```

Step 5: Add a file `.cargo/config` under tlsn-extension project root with the following content, as wasm-bindgen-rayon [suggested](https://github.com/GoogleChromeLabs/wasm-bindgen-rayon#using-config-files).

```bash
[target.wasm32-unknown-unknown]
rustflags = ["-C", "target-feature=+atomics,+bulk-memory,+mutable-globals"]

[unstable]
build-std = ["panic_abort", "std"]
```

Step 6: Build the wasm and run the dev server
```bash
yarn build-and-start
```
Step 7: Open the browser (I've been using Chrome), visit `http://localhost:8080/`, and open the console. You should see 

![image](https://github.com/tlsnotary/tlsn-extension/assets/8223657/d48ee0a1-8fa1-45e3-b398-316b5f984a8b)

and some logs, and then

![image](https://github.com/tlsnotary/tlsn-extension/assets/8223657/ca0ac8ab-6430-4d4b-9e37-ba491a230bf4)

You shouldn't be able to see `!@# 3` in the console.

**Comment by mhchia** [comment]
Seems to do with websocket. 🤔 With nightly + rustflags, writing to websocket results in an error 

![image](https://github.com/tlsnotary/tlsn-extension/assets/8223657/beaabd91-3165-44d9-b5c5-ccaa24e1c771)

**Comment by mhchia** [comment]
With nightly + rustflags, reading and writing to the websocket results in an error.

## Steps to Reproduce
Step 1: Checkout to the branch `mhchia/debugging-nightly-atomics-websocket`
```
git fetch origin
git checkout mhchia/debugging-nightly-atomics-websocket
```
Step 2: Ensure `.cargo/config` has the content
```
[target.wasm32-unknown-unknown]
rustflags = ["-C", "target-feature=+atomics,+bulk-memory,+mutable-globals"]

[unstable]
build-std = ["panic_abort", "std"]
```
Step 3: Build and run. You should see "Error: Failed to write to websocket" in the console
![image](https://github.com/tlsnotary/tlsn-extension/assets/8223657/6bb134e2-3461-44a0-b219-a828b3ce4cf5)

**Comment by mhchia** [comment]
I've made a smaller example to reproduce: https://github.com/mhchia/test-ws-stream-wasm

**Comment by mhchia** [comment]
Worked around by increasing wasm memory to 4GB. Will have another issue elaborate more on this

**Comment by mhchia** [comment]
## Websocket error workaround
Shout out again to @themighty1 for locating the issue in ws-stream-wasm and having a patch. With the patch, tlsn-extension can run without websocket errors, until prover is finalizing. Then, I encountered an error which looks an out-of-memory issue.

![image](https://github.com/tlsnotary/tlsn-extension/assets/8223657/a0f7c0b8-ab17-4a46-9018-27650add3af4)

I workaround it by increasing wasm memory to 4GB (4x the default value), and prover can work successfully.

## Rayon

By adding back `rayon` and `wasm-bindgen-rayon`, tlsn-extension can run with multi-threads.

![image](https://github.com/tlsnotary/tlsn-extension/assets/8223657/753c0564-cbf0-40c7-9210-9e6b6bf7dc8a)

With 6~12 threads on my laptop (`hardwareConcurrency=12`) , it takes ~2 mins, 3x faster than single-threaded (6 mins).

![image](https://github.com/tlsnotary/tlsn-extension/assets/8223657/b837da8e-76c0-46e0-b4dd-9b755d745b25)

## Steps to reproduce
Step 0: Run notary and websockify as https://github.com/tlsnotary/tlsn-extension/issues/6#issue-1878436657

Step 1: For extension, checkout to the branch `mhchia/workaround-ws-stream-wasm-and-work-with-rayon`
```
git fetch origin
git checkout mhchia/workaround-ws-stream-wasm-and-work-with-rayon
```

Step 2: Ensure you're using nightly rust
```
echo "nightly-2022-12-12" > rust-toolchain
```

Step 3: Change wasm memory to 4GB. Copy the following content and paste to `.cargo/config`
```
[target.wasm32-unknown-unknown]
rustflags = [
    "-C",
    "target-feature=+atomics,+bulk-memory,+mutable-globals",
    "-C",
    "link-arg=--max-memory=4294967296"
]

[unstable]
build-std = ["panic_abort", "std"]
```
Step 4: Build and run
```bash
yarn build-and-start
```

You might see some warnings but that's fine
![](https://hackmd.io/_uploads/BJpr7h303.png)

Step 5: Open the browser (I've been using Chrome), visit http://localhost:8080/, and open the console. It should run successfully.

## Concerns
This workaround only unblocks us to test if rayon can work in browser. Should try to find a better solution without increasing the memory, for example, a patch without oom happening.

**Comment by mhchia** [comment]
Now extension works with nightly and atomics. OOM issue is tracked in #13 .
